helix-core/src/test/java/org/apache/helix/integration/rebalancer/WagedRebalancer/TestWagedRebalanceHardConstraint.java

```diff
@@ -0,0 +1,286 @@
+      _controller.syncStop();
+    }
+    for (MockParticipantManager p : _participants) {
+      if (p != null && p.isConnected()) {
+        p.syncStop();
+      }
+    }
+    deleteCluster(CLUSTER_NAME);
+  }
+
+  private void validate() {
+    HelixClusterVerifier _clusterVerifier =
+        new StrictMatchExternalViewVerifier.Builder(CLUSTER_NAME).setZkAddr(ZK_ADDR)
+            .setDeactivatedNodeAwareness(true).setResources(_allDBs)
+            .setWaitTillVerify(TestHelper.DEFAULT_REBALANCE_PROCESSING_WAIT_TIME)
+            .build();
+    try {
+      Assert.assertTrue(_clusterVerifier.verify(5000));
+    } finally {
+      _clusterVerifier.close();
+    }
+    for (String db : _allDBs) {
+      IdealState is =
+          _gSetupTool.getClusterManagementTool().getResourceIdealState(CLUSTER_NAME, db);
+      ExternalView ev =
+          _gSetupTool.getClusterManagementTool().getResourceExternalView(CLUSTER_NAME, db);
+      validateIsolation(is, ev);
+    }
+  }
+  private void printCurrentState() {
+    List<String> instances =
+        _gSetupTool.getClusterManagementTool().getInstancesInCluster(CLUSTER_NAME);
+    HelixDataAccessor dataAccessor = new ZKHelixDataAccessor(CLUSTER_NAME, _baseAccessor);
+
+    for (String instance : instances) {
+      LiveInstance liveInstance =
+          dataAccessor.getProperty(dataAccessor.keyBuilder().liveInstance(instance));
+      String sessionId = liveInstance.getEphemeralOwner();
+      List<CurrentState> currentStates = dataAccessor.getChildValues(dataAccessor.keyBuilder().currentStates(instance, sessionId),
+          true);
+      LOG.info("\n\nCurrentState for instance: " + instance);
+      for (CurrentState currentState : currentStates) {
+          LOG.info("\t" + currentState.getRecord().getMapFields().keySet().toString());
+      }
+    }
+  }
+  private void validateIdealState(boolean forAll, String onlyResource) {
+    for (String db : _allDBs) {
+      IdealState is =
+            _gSetupTool.getClusterManagementTool().getResourceIdealState(CLUSTER_NAME, db);
+      IdealState prevIdealState = _prevIdealState.get(db);
+      if (forAll && prevIdealState != null) {
+        Assert.assertNotSame(is.getRecord().getMapFields(), prevIdealState.getRecord().getMapFields());
+      } else if (prevIdealState != null && onlyResource.equals(db)) {
+        Assert.assertNotSame(is.getRecord().getMapFields(), prevIdealState.getRecord().getMapFields());
+      }
+      LOG.info("IdealState for resource: " + db);
+      for (String partition : is.getPartitionSet()) {
+          Map<String, String> assignmentMap = is.getRecord().getMapField(partition);
+          LOG.info("\tFor partition: " + partition);
+          for (String instance : assignmentMap.keySet()) {
+            LOG.info("\t\tinstance: " + instance);
+          }
+      }
+      _prevIdealState.put(db, is);
+    }
+  }
+  private void validateIsolation(IdealState is, ExternalView ev) {
```

Comment:
What is the purpose of this? Do you want to make sure 3 replicas are all alive at the same time? Or you are trying to confirm the N -> N+1 -> N is done?

Reply:
the preferred list in IS would have changed, indicating that the partitions moved from one host to other and when that happens, the rules are broken.